### PR TITLE
add verification to gluon dataset

### PR DESCRIPTION
**Pull request by szha**
@piiswrong

### CODE REVIEW

File: python/mxnet/gluon/utils.py [hunk 1]
```diff
@@ -144,7 +174,7 @@ def download(url, path=None, overwrite=False):
     else:
         fname = path
 
-    if overwrite or not os.path.exists(fname):
+    if overwrite or not os.path.exists(fname) or not sha1hex or not verified_sha1(fname, sha1hex):
```
Comment (piiswrong) [inline python/mxnet/gluon/utils.py]: this means always overwrite when sha1hex is None

File: python/mxnet/gluon/utils.py [hunk 2]
```diff
@@ -119,7 +120,33 @@ def _indent(s_, numSpaces):
...
+
+    Returns
+    -------
+    bool
+        Whether the file content matches the expected hash.
+    """
+    sha1 = hashlib.sha1()
+    with open(file_path, 'rb') as f:
+        while True:
+            data = f.read(1048576)
```
Comment (piiswrong) [inline python/mxnet/gluon/utils.py]: where does 1048576 come from

Comment (szha) [inline python/mxnet/gluon/utils.py]: This is 1mb.

File: python/mxnet/gluon/utils.py [hunk 3]
```diff
@@ -119,7 +120,33 @@ def _indent(s_, numSpaces):
     return s
 
 
-def download(url, path=None, overwrite=False):
+def verified_sha1(file_path, sha1hex):
```
Comment (piiswrong) [inline python/mxnet/gluon/utils.py]: check_sha1(filename, sha1_hash)

File: python/mxnet/gluon/utils.py [hunk 4]
```diff
@@ -119,7 +120,33 @@ def _indent(s_, numSpaces):
...
+        while True:
+            data = f.read(1048576)
+            if not data:
+                break
+            sha1.update(data)
+
+    return sha1.hexdigest() == sha1hex
+
+
+def download(url, path=None, overwrite=False, sha1hex=None):
```
Comment (piiswrong) [inline python/mxnet/gluon/utils.py]: sha1hex -> sha1_hash

Comment (szha) [inline python/mxnet/gluon/utils.py]: a hash can be in any form (bytes, long long, string with different radix). hex there is to signify what the input should be (i.e. hexadecimal string).

**Comment by piiswrong** [comment]
Please add tests for mnist and cifar10

**Pull request merged**